### PR TITLE
🟡(change): cache organization response

### PR DESCRIPTION
**Pull request by ruancomelli**
Organization is pulled quite a bit from the application, but should not be updated often.

## Summary by Sourcery

This pull request introduces caching for organization data using Redis, removes the rainbow animation feature from the theme provider, and consolidates organization-related API endpoints and hooks. Additionally, it updates dependencies to include the new Redis client library.

- **New Features**:
    - Introduced caching for organization data using Redis to reduce database load and improve response times.
- **Enhancements**:
    - Removed the rainbow animation feature from the theme provider and related components.
    - Consolidated organization-related API endpoints and hooks for better consistency and maintainability.
- **Chores**:
    - Updated dependencies in go.mod to include new Redis client library.

### CODE REVIEW

**Comment by ruancomelli** [comment]
This is an experiment review for experiment `improve_review_guide`
Run ID: `improve_review_guide/run_2024-08-12T01-08-43_v1-21-0-74-g8aa259cc3-dirty`
The benchmark review for this pull request can be found at https://github.com/sourcery-ai-experiments/Trenova/pull/19.
This pull request was cloned from `https://github.com/sourcery-ai-experiments/Trenova/pull/11`. _(Note: the URL is not a link to avoid triggering a notification on the original pull request.)_

<details>
<summary>Experiment configuration</summary>

```yaml
review_config:
  # User configuration for the review
  # - benchmark - use the user config from the benchmark reviews
  # - <value> - use the value directly
  user_review_config:
    enable_ai_review: true
    enable_rule_comments: false

    enable_complexity_comments: false
    enable_security_comments: false
    enable_tests_comments: false
    enable_comment_suggestions: false

    enable_claude_comments: true
    enable_functionality_review: false

    enable_pull_request_summary: false
    enable_review_guide: true

    enable_approvals: false

  ai_review_config:
    # The model responses to use for the experiment
    # - benchmark - use the model responses from the benchmark reviews
    # - llm - call the language model to generate responses
    model_responses:
      comments_model: benchmark
      comment_area_model: benchmark
      comment_validation_model: benchmark
      comment_suggestion_model: benchmark
      complexity_model: benchmark
      docstrings_model: benchmark
      functionality_model: benchmark
      overall_comments_model: benchmark
      security_model: benchmark
      tests_model: benchmark
      pull_request_summary_model: benchmark
      review_guide_model: llm

# The pull request dataset to run the experiment on
pull_request_dataset:
  # open-source PRs
# - https://github.com/sourcery-ai-experiments/gpt-crawler/pull/1
# - https://github.com/sourcery-ai-experiments/graphviz2drawio/pull/1
# - https://github.com/sourcery-ai-experiments/graphviz2drawio/pull/2
# - https://github.com/sourcery-ai-experiments/iptux/pull/5
# - https://github.com/sourcery-ai-experiments/iptux/pull/6
# - https://github.com/sourcery-ai-experiments/kalima/pull/3
# - https://github.com/sourcery-ai-experiments/pydantic_tfl_api/pull/1
# - https://github.com/sourcery-ai-experiments/pydantic_tfl_api/pull/2
# - https://github.com/sourcery-ai-experiments/Trenova/pull/11
#   # Sourcery PRs
# - https://github.com/sourcery-ai/core/pull/4055
# - https://github.com/sourcery-ai/core/pull/4263
# - https://github.com/sourcery-ai/core/pull/4352
# - https://github.com/sourcery-ai/core/pull/4364
# - https://github.com/sourcery-ai/core/pull/4366
# - https://github.com/sourcery-ai/core/pull/4390
# - https://github.com/sourcery-ai/core/pull/4402
# - https://github.com/sourcery-ai/core/pull/4445
# - https://github.com/sourcery-ai/core/pull/4462
- https://github.com/sourcery-ai/core/pull/4463

review_comment_labels: []

# Benchmark reviews generated by running
#   python -m scripts.experiment benchmark <experiment_name>
benchmark_reviews:
- dataset_pull_request: https://github.com/sourcery-ai/core/pull/4463
  review_pull_request: https://github.com/sourcery-ai-experiments/core/pull/236
- dataset_pull_request: https://github.com/sourcery-ai-experiments/gpt-crawler/pull/1
  review_pull_request: https://github.com/sourcery-ai-experiments/gpt-crawler/pull/10
- dataset_pull_request: https://github.com/sourcery-ai-experiments/iptux/pull/5
  review_pull_request: https://github.com/sourcery-ai-experiments/iptux/pull/32
- dataset_pull_request: https://github.com/sourcery-ai-experiments/iptux/pull/6
  review_pull_request: https://github.com/sourcery-ai-experiments/iptux/pull/33
- dataset_pull_request: https://github.com/sourcery-ai-experiments/kalima/pull/3
  review_pull_request: https://github.com/sourcery-ai-experiments/kalima/pull/17
- dataset_pull_request: https://github.com/sourcery-ai/core/pull/4055
  review_pull_request: https://github.com/sourcery-ai-experiments/core/pull/258
- dataset_pull_request: https://github.com/sourcery-ai/core/pull/4263
  review_pull_request: https://github.com/sourcery-ai-experiments/core/pull/259
- dataset_pull_request: https://github.com/sourcery-ai/core/pull/4352
  review_pull_request: https://github.com/sourcery-ai-experiments/core/pull/260
- dataset_pull_request: https://github.com/sourcery-ai/core/pull/4364
  review_pull_request: https://github.com/sourcery-ai-experiments/core/pull/261
- dataset_pull_request: https://github.com/sourcery-ai/core/pull/4366
  review_pull_request: https://github.com/sourcery-ai-experiments/core/pull/262
- dataset_pull_request: https://github.com/sourcery-ai/core/pull/4390
  review_pull_request: https://github.com/sourcery-ai-experiments/core/pull/263
- dataset_pull_request: https://github.com/sourcery-ai/core/pull/4402
  review_pull_request: https://github.com/sourcery-ai-experiments/core/pull/264
- dataset_pull_request: https://github.com/sourcery-ai/core/pull/4445
  review_pull_request: https://github.com/sourcery-ai-experiments/core/pull/265
- dataset_pull_request: https://github.com/sourcery-ai/core/pull/4462
  review_pull_request: https://github.com/sourcery-ai-experiments/core/pull/266
- dataset_pull_request: https://github.com/sourcery-ai-experiments/graphviz2drawio/pull/1
  review_pull_request: https://github.com/sourcery-ai-experiments/graphviz2drawio/pull/23
- dataset_pull_request: https://github.com/sourcery-ai-experiments/graphviz2drawio/pull/2
  review_pull_request: https://github.com/sourcery-ai-experiments/graphviz2drawio/pull/24
- dataset_pull_request: https://github.com/sourcery-ai-experiments/pydantic_tfl_api/pull/1
  review_pull_request: https://github.com/sourcery-ai-experiments/pydantic_tfl_api/pull/19
- dataset_pull_request: https://github.com/sourcery-ai-experiments/pydantic_tfl_api/pull/2
  review_pull_request: https://github.com/sourcery-ai-experiments/pydantic_tfl_api/pull/20
- dataset_pull_request: https://github.com/sourcery-ai-experiments/Trenova/pull/11
  review_pull_request: https://github.com/sourcery-ai-experiments/Trenova/pull/19

```

</details>

**Comment by SourceryAI** [comment]
## Reviewer's Guide by Sourcery

This pull request introduces caching for organization data using Redis to improve performance and reduce database load. It also consolidates organization-related API endpoints and hooks, removes the rainbow animation feature from the theme provider, and updates dependencies to include the new Redis client library.

The main changes include:

1. Implementation of Redis caching for organization data in the OrganizationService.
2. Consolidation of organization-related API endpoints and hooks in the frontend.
3. Removal of the rainbow animation feature from the theme provider and related components.
4. Addition of a new Redis client package and integration with the server.
5. Updates to the server configuration to include Redis cache options.
6. Modifications to the Organization model to support serialization for caching.

These changes aim to improve the application's performance by reducing database queries for frequently accessed organization data and streamlining the codebase by removing unused features and consolidating related functionality.

### File-Level Changes

| Files | Changes |
| ----- | ------- |
| `internal/api/services/organization.service.go` | Implemented Redis caching for organization data in the OrganizationService |
| `web/frontend/src/hooks/useQueries.ts`<br/>`web/frontend/src/services/OrganizationRequestService.ts` | Consolidated organization-related API endpoints and hooks in the frontend |
| `web/frontend/src/components/ui/theme-provider.tsx`<br/>`web/frontend/src/components/layout/user-avatar-menu.tsx`<br/>`web/frontend/src/components/layout/topbar.tsx` | Removed the rainbow animation feature from the theme provider and related components |
| `pkg/redis/client.go`<br/>`internal/server/server.go`<br/>`cmd/serve.go` | Added a new Redis client package and integrated it with the server |
| `config/server_config.go` | Updated server configuration to include Redis cache options |
| `pkg/models/organization.go` | Modified the Organization model to support serialization for caching |
| `go.mod` | Updated dependencies in go.mod to include new Redis client library |

---

<details>
<summary>Tips</summary>

- Trigger a new Sourcery review by commenting `@sourcery-ai review` on the pull request.
- Continue your discussion with Sourcery by replying directly to review comments.
- You can change your review settings at any time by accessing your [dashboard](https://app.sourcery.ai):
    - Enable or disable the Sourcery-generated pull request summary or reviewer's guide;
    - Change the review language;
- You can always [contact us](mailto:support@sourcery.ai) if you have any questions or feedback.

</details>

File: internal/server/server.go [hunk 1]
```diff
@@ -160,6 +166,23 @@ func (s *Server) InitDB() *bun.DB {
 	return s.DB
 }
 
+func (s *Server) InitCache() *redis.Client {
```
Comment (SourceryAI) [inline internal/server/server.go:169]: **suggestion:** Make cache client timeout configurable.

The cache client is being initialized with a fixed timeout of 5 seconds. Consider making this timeout configurable to provide more flexibility and allow for better tuning based on different environments.

File: pkg/redis/client.go [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+	if err := client.initialize(); err != nil {
+		logger.Fatal().Err(err).Msg("Failed to initialize redis client")
+		return nil
+	}
+
+	return client
+}
+
+func (r *Client) initialize() error {
```
Comment (SourceryAI) [inline pkg/redis/client.go:30]: **suggestion:** Use a more graceful error handling approach for Redis client initialization.

Logging a fatal error and terminating the application if the Redis client initialization fails might not be the best approach. Consider implementing a more graceful error handling mechanism to handle initialization failures.

**Review by SourceryAI — COMMENTED** [review]
Hey @ruancomelli - I've reviewed your changes - here's some feedback:

**Overall Comments**:
- Consider encapsulating repetitive error logging patterns into helper functions to reduce code duplication.
- Ensure that the removal of the rainbow animation feature is well-documented in the release notes or migration guide for users who might be relying on that feature.
- Fix the typo in the comment: 'Try toi fetch the organization from the cache' should be 'Try to fetch the organization from the cache'.

<details>
<summary>Here's what I looked at during the review</summary>

- 🟡 **General issues**: 2 issues found
- 🟢 **Documentation**: all looks good
</details>

***

[LangSmith trace](https://smith.langchain.com/o/7d8518ad-6dd3-5c7b-96d2-757769296335/projects/p/88560225-5072-48bc-967c-dc54da53f265/r/9b2a738d-8cf5-4c69-8347-5e5ebc921120?poll=true)

<sub>
Help me be more useful! Please click 👍 or 👎 on each comment to tell me if it was helpful.
</sub>

**Pull request closed without merging**